assets/js/phoenix_live_view.js

```diff
@@ -214,6 +213,7 @@ export class LiveSocket {
     this.viewLogger = opts.viewLogger
     this.activeElement = null
     this.prevActive = null
+    this.document = document;
```

Comment:
@raygesualdo Thoughts here?  I needed this b/c we needed a way to use the jsdom document and assign it to `liveSocket.document` in a test.  Know of another way of doing this?

Reply:
This could be handled by overriding `global.document` for each test. The below example is pseudo-code and I can't guarantee it works as written, but this is generally how you could implement it (using the `displayError` test in `view_test.js`:

```javascript
test('displayError', async () => {
  const originalDocument = global.document
  global.document = dom().window.document

  let liveSocket = new LiveSocket('/live')
  let loader = document.createElement('span')
  let phxView = document.querySelector('[data-phx-view]')
  phxView.parentNode.insertBefore(loader, phxView.nextSibling)
  let el = document.querySelector('[data-phx-view]')

  let view = new View(el, liveSocket)
  view.displayError()
  expect(el.classList.contains('phx-disconnected')).toBeTruthy()
  expect(el.classList.contains('phx-error')).toBeTruthy()
  expect(loader.style.display).toEqual('block')

  global.document = originalDocument
})
```

If we were to go this route, we would want to move this to `beforeEach` and `afterEach` handlers.

Reply:
Which reminds me, we probably want to group the tests in each file with at least one level of `describe` block if not two. See `browser_test.js` for what this pattern looks like.

Run `npm test` to see how the test output differs across the test files.

Reply:
@raygesualdo If I do a simple test like this, it doesn't seem to be propagating the document to the `live_view.js` code. I would assume `document` (when queried inside the production code) should be this, but it doesn't seem to be.

```
  beforeEach(() => {
    global.document = dom().window.document;
  });

  test('connect', async () => {
    let liveSocket = new LiveSocket('/live');

    let socket = liveSocket.connect();

    expect(liveSocket.views['container'].destroy).toBeDefined();
    // i'm not sure why this is
    expect(socket).toBeUndefined();
  });
```

Reply:
Ahh, turns out it has to be overridden in a slightly different way (certain globals are more protected than others). I tested it this time 😁 and the following snippet works:

```javascript
  let originalDocument;
  
  beforeEach(() => {
    if (!originalDocument) originalDocument = global.document;
    Object.defineProperty(global, "document", {
      value: dom().window.document
    });
  });

  afterAll(() => {
    Object.defineProperty(global, "document", {
      value: originalDocument
    });
  });

  test("connect", async () => {
    let liveSocket = new LiveSocket("/live");
    let socket = liveSocket.connect();

    expect(liveSocket.views["container"].destroy).toBeDefined();
    // i'm not sure why this is
    expect(socket).toBeUndefined();
  });
```

It's a little more verbose but doing the same thing: mocking `document` before each test then resetting back to the original at the end.

Reply:
Ok next issue 😆.   Getting this to work on travis.

https://travis-ci.org/phoenixframework/phoenix_live_view/jobs/508722161

Reply:
Do all the tests pass locally?

Reply:
I'll try to take a look first thing in the morning.

Reply:
Ah its node 8.  Jest is bleeding edge :)